cli/test_routing.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2018 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/prometheus/alertmanager/client"
+	"github.com/prometheus/alertmanager/dispatch"
+	"github.com/xlab/treeprint"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+const routingTestHelp = `Test alert routing
+
+Will return receiver names which the alert with given labels resolves to.
+(If resolves to multiple receivers, they are printed out joined using ',' in order as defined in the routing tree)
+
+Routing is loaded from configuration file or remote Alertmanager.
+Use --config.file or --alertmanager.url flags to specify where to look for the configuration.
+Than specify labels defining the alert.
+
+Example:
+
+./amtool config routes test --config.file=doc/examples/simple.yml --verify.receivers=team-DB-pager service=database
+
+`
+
+func configureRoutingTestCmd(cc *kingpin.CmdClause, c *routingShow) {
+	var routingTestCmd = cc.Command("test", routingTestHelp)
+
+	routingTestCmd.Flag("verify.receivers", "Checks if specified receivers matches resolved receivers. (Returns exit status 1 if not)").StringVar(&c.expectedReceivers)
+	routingTestCmd.Flag("tree", "Prints out matching routes tree").BoolVar(&c.debugTree)
+	routingTestCmd.Arg("labels", "Alert labels in name=value notation").StringsVar(&c.labels)
+	routingTestCmd.Action(execWithTimeout(c.routingTestAction))
+}
+
+// resolveAlertReceivers returns list of receiver names which given LabelSet resolves to.
+func resolveAlertReceivers(mainRoute *dispatch.Route, labels *client.LabelSet) ([]string, error) {
+	var (
+		finalRoutes []*dispatch.Route
+		receivers   []string
+	)
+	finalRoutes = mainRoute.Match(ConvertCientToCommonLabelSet(*labels))
+	for _, r := range finalRoutes {
+		receivers = append(receivers, r.RouteOpts.Receiver)
+	}
+	return receivers, nil
+}
+
+func printMatchingTree(mainRoute *dispatch.Route, ls client.LabelSet) {
+	tree := treeprint.New()
+	getMatchingTree(mainRoute, tree, ls)
+	fmt.Println("Matching routes:")
+	fmt.Println(tree.String())
+	fmt.Print("\n")
+}
+
+func (c *routingShow) routingTestAction(ctx context.Context, _ *kingpin.ParseContext) error {
+
+	// Load configuration form file or URL.
```

Comment:
I don't think this comment is necessary, I think the next line is pretty self-documenting